hazelcast/src/test/java/com/hazelcast/jet/JobListenerTest.java

```diff
@@ -0,0 +1,387 @@
+                            "Jet: STARTING -> RUNNING",
+                            "Jet: RUNNING -> FAILED (" + failure + ")");
+                });
+    }
+
+    @Test
+    public void testListener_restartOnException() {
+        testListener(new JobConfig().setAutoScaling(true),
+                TestSources.itemStream(1,
+                        (t, s) -> {
+                            if (s == 2) {
+                                throw new RestartableException();
+                            }
+                            return s;
+                        }),
+                (job, listener) -> {
+                    assertEqualsEventually(listener::runCount, 2);
+                    job.cancel();
+                    assertThrows(CancellationException.class, job::join);
+                    assertTailEquals(listener.log.stream().map(SIMPLIFY).collect(toList()),
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "Jet: RUNNING -> NOT_RUNNING (com.hazelcast.jet.RestartableException)",
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "User: RUNNING -> FAILED (CANCEL)");
+                });
+    }
+
+    @Test
+    public void testListener_suspendOnFailure() {
+        testListener(new JobConfig().setSuspendOnFailure(true),
+                TestSources.itemStream(1, (t, s) -> 1 / (2 - s)),
+                (job, listener) -> {
+                    assertJobStatusEventually(job, SUSPENDED);
+                    String failure = job.getSuspensionCause().errorCause().split("\n", 3)[1];
+                    job.cancel();
+                    assertThrows(CancellationException.class, job::join);
+                    assertTailEquals(listener.log,
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING",
+                            "Jet: RUNNING -> SUSPENDED (" + failure + ")",
+                            "User: SUSPENDED -> FAILED (CANCEL)");
+                });
+    }
+
+    @Test
+    public void testListenerDeregistration() {
+        testListener(TestSources.itemStream(1, (t, s) -> s),
+                (job, listener) -> {
+                    assertJobStatusEventually(job, RUNNING);
+                    instance.get().getJet().removeJobStatusListener(listener);
+                    job.cancel();
+                    assertThrows(CancellationException.class, job::join);
+                    assertTailEquals(listener.log,
+                            "Jet: NOT_RUNNING -> STARTING",
+                            "Jet: STARTING -> RUNNING");
+                });
+    }
+
+    @Test
+    public void testLightListener_waitForCompletion() {
+        testLightListener(finiteStream(1000, 2),
+                job -> {
+                    job.join();
+                    sleepSeconds(1);
```

Comment:
This can be removed by changing the order of `jobCompletionFuture.complete` and `jobService.publishEvent`
calls in `LightMasterContext.finalizeJob`. However, it would cause propagating `COMPLETED`/`FAILED`
statuses before they are actually set, i.e. in the listener, `Job.getStatus()` will return `RUNNING`.